packages/mui-material/src/Button/Button.js

```diff
@@ -343,8 +391,118 @@ const ButtonEndIcon = styled('span', {
         marginRight: -2,
       },
     },
+    {
+      props: { loadingPosition: 'end', loading: true },
+      style: {
+        transition: theme.transitions.create(['opacity'], {
+          duration: theme.transitions.duration.short,
+        }),
+        opacity: 0,
+      },
+    },
+    {
+      props: { loadingPosition: 'end', loading: true, fullWidth: true },
+      style: {
+        marginLeft: -8,
+      },
+    },
     ...commonIconStyles,
   ],
+}));
+
+const ButtonLoadingIndicator = styled('span', {
+  name: 'MuiButton',
+  slot: 'LoadingIndicator',
+  overridesResolver: (props, styles) => styles.loadingIndicator,
+})(({ theme }) => ({
+  display: 'none',
+  position: 'absolute',
+  visibility: 'visible',
+  variants: [
+    { props: { loading: true }, style: { display: 'flex' } },
+    {
+      props: {
+        loadingPosition: 'start',
+        size: 'small',
+      },
+      style: {
+        left: 10,
+      },
+    },
+    {
+      props: ({ loadingPosition, size }) => loadingPosition === 'start' && size !== 'small',
+      style: {
+        left: 14,
+      },
+    },
```

Comment:
Would applying the base styles first work?

```js
    {
      props: { loadingPosition: 'start' },
      style: {
        left: 14,
      },
    },
    {
      props: {
        loadingPosition: 'start',
        size: 'small',
      },
      style: {
        left: 10,
      },
    },
```

I think this would make it more readable and maintainable. What do you think?

(The same applies to the other `'small'` size styles applied elsewhere in this PR.